src/meili/candidate_search.cc

```diff
@@ -109,7 +109,10 @@ CandidateQuery::WithinSquaredDistance(const midgard::PointLL& location,
 
     if (correlated.edges.size()) {
       // Add back if it is an edge correlated or it's a node correlated
-      // but it's not added yet
+      // but it's not added yet. By only allow one snapped_node per intersection match
```

Comment:
This is where the intersection mismatching starts. The alternative fix here is to remove the condition . Instead of only allowing one snapped_node (one edge candidate for the intersection), we add all of edges ends/starts on that node to the candidates. And later, the use an exclusive search from the previous state and the next state to match edge transition consistency.

Reply:
This is the important part which as two competing ideas going on. That `|| visited_nodes.insert(snapped_node).second` will exclude certain edge candidates at the node (randomly, first one wins). Lets take the following example:

```
         0
         ||
       C || D
         ||
         ||
   A     ||     E
======== 1 ======== 2
   B     ||     F
         ||
         ||
       G || H
         ||
```

The input locations are labled `0` through `2`, the edges are labeled `A` through `H`. Imagine that input point `1` is at the node where all the edges cross. But the candidate_search encounters edges `A` and `B` first. What on earth can the map matching be? If edges `A` `C` `E` and `H` end at the node at `1`, then you'd get something like: `C.0->A.1->F.1`. However `C` does not connect to `A` so what do we have, a discontinuity?

The problem here is we want to minimize the number of candidates that we have in any given state of the HMM because it exploses the permutations we have to evaluate. To the router though `A.1` is just as good as `F.0`. That is why this `if` check excludes adding more candidates.